lib/Runtime/Base/FunctionBody.cpp

```diff
@@ -55,6 +56,213 @@ namespace Js
+        }
+        else
+        {
+            Assert(false);
+        }
+    }
+
+    template<class T, typename FIELDS>
+    void AuxPtrs<T, FIELDS>::allocAuxPtr(T* _this, uint8 count, Recycler* recycler)
+    {
+        Assert(count >= AuxPtrs32::MaxCount);
+        auto requestSize = sizeof(AuxPtrs<T, FIELDS>) + (count - 1)*sizeof(void*);
+        auto allocSize = ::Math::Align<uint8>((uint8)requestSize, 16);
+        auto capacity = (uint8)((allocSize - offsetof(AuxPtrsT, ptrs)) / sizeof(void*));
+
+        if (_this->auxPtrs->count != AuxPtrs32::MaxCount) // expanding
+        {
+            _this->auxPtrs = RecyclerNewWithBarrierPlusZ(recycler, allocSize - sizeof(AuxPtrsT), AuxPtrsT, capacity, _this->auxPtrs);
+        }
+        else // promoting
+        {
+            _this->auxPtrs = RecyclerNewWithBarrierPlusZ(recycler, allocSize - sizeof(AuxPtrsT), AuxPtrsT, capacity, (AuxPtrs32*)(void*)_this->auxPtrs);
+        }
+    }
+
+    //static void* get_AuxPtr(T* _this, FIELDS e, Recycler* recycler) const;
+    //static void set_AuxPtr(T* _this, FIELDS e, void* ptr, Recycler* recycler);
+    template<class T, typename FIELDS>
+    inline void* AuxPtrs<T, FIELDS>::get_AuxPtr(const T* _this, FIELDS e)
+    {
+        if (_this->auxPtrs == nullptr)
+        {
+            return nullptr;
+        }
+        if (_this->auxPtrs->count == AuxPtrs16::MaxCount)
+        {
+            return ((AuxPtrs16*)(void*)_this->auxPtrs)->get(e);
+        }
+        if (_this->auxPtrs->count == AuxPtrs32::MaxCount)
+        {
+            return ((AuxPtrs32*)(void*)_this->auxPtrs)->get(e);
+        }
+        return _this->auxPtrs->get(e);
+    }
+    template<class T, typename FIELDS>
+    void AuxPtrs<T, FIELDS>::set_AuxPtr(T* _this, FIELDS e, void* ptr, Recycler* recycler)
+    {
+        if (ptr == nullptr && get_AuxPtr(_this, e) == nullptr)
+        {
+            return;
+        }
+        if (_this->auxPtrs == nullptr)
+        {
+            AuxPtrs<FunctionProxy, FIELDS>::allocAuxPtrFix(_this, 16, recycler);
+            bool ret = ((AuxPtrs16*)(void*)_this->auxPtrs)->set(e, ptr);
+            Assert(ret);
+            return;
+        }
+        if (_this->auxPtrs->count == AuxPtrs16::MaxCount)
+        {
+            bool ret = ((AuxPtrs16*)(void*)_this->auxPtrs)->set(e, ptr);
+            if (ret)
+            {
+                return;
+            }
+            else
+            {
+                AuxPtrs<FunctionProxy, FIELDS>::allocAuxPtrFix(_this, 32, recycler);
+            }
+        }
+        if (_this->auxPtrs->count == AuxPtrs32::MaxCount)
+        {
+            bool ret = ((AuxPtrs32*)(void*)_this->auxPtrs)->set(e, ptr);
+            if (ret)
+            {
+                return;
+            }
+            else
+            {
+                AuxPtrs<FunctionProxy, FIELDS>::allocAuxPtr(_this, AuxPtrs32::MaxCount + 1, recycler);
```

Comment:
Are you talking about names or capacity?
AuxPtr16, use 16 bytes, it’s 8 bytes metadata+ 8byte pointers(1 on x64 or 2 on x86)
AuxPtr32, use 32 bytes, it’s 8 bytes metadata+ 24byte pointers(3 on x64 or 6 on x86)
These two are fixed structures, and using linear search to locate the pointers because the position is not saved in the metadata

AuxPtrs is an dynamic expanding structure, when AuxPtr32 can’t hold that many pointers, it will promote to AuxPtrs.
AuxPtrs use 24 bytes metadata(for now), it saves position of each instantiated pointer so it does not use linear search.

From a statistic, about 70% will have the auxiliary pointers structure instantiated. Among on amd64, 25% has only one pointer instantiated and hold in AuxPtr16, another 50% has 2 or 3 pointers, will hold in AuxPtr16 and them promote to AuxPtr32, the left over 25% has 4 or more pointers instantiated and will promote to AuxPtrs

Thanks,
Lei



From: Abhijith Chatra [mailto:notifications@github.com]
Sent: Friday, February 5, 2016 6:51 AM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Lei Shi <leish@microsoft.com>
Subject: Re: [ChakraCore] Move rarely used pointers from FunctionBody to a dynamic structure to reduce FunctionBody size (#268)


In lib/Runtime/Base/FunctionBody.cpp<https://github.com/Microsoft/ChakraCore/pull/268#discussion_r52023895>:

> +            }

> +            else

> +            {

> +                AuxPtrs<FunctionProxy, FIELDS>::allocAuxPtrFix(_this, 32, recycler);

> +            }

> +        }

> +        if (_this->auxPtrs->count == AuxPtrs32::MaxCount)

> +        {

> +            bool ret = ((AuxPtrs32*)(void*)_this->auxPtrs)->set(e, ptr);

> +            if (ret)

> +            {

> +                return;

> +            }

> +            else

> +            {

> +                AuxPtrs<FunctionProxy, FIELDS>::allocAuxPtr(_this, AuxPtrs32::MaxCount + 1, recycler);

Got it. Very cool!. I have one question, how did you make the choice of AuxPtrs16 & AuxPtrs32? Why not AuxPtrs32 & AuxPtrs64 for x64?

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/268/files#r52023895>.